23.1.2+ (#1)
friedcucumber/maui-mail-merge: README.md

# Mail Merge

> This example shows you how to send a templated email to a contact with his/her name and data. Mail merge is included in our [Office File API](https://www.devexpress.com/products/net/office-file-api/) subscription.

Reviewer: ```suggestion
This example uses Mail Merge to send emails. Data placeholders in an email template obtain their values from database records. Mail Merge functionality is included in our [Office File API](https://www.devexpress.com/products/net/office-file-api/) subscription.
```

## Requirements

* Please register the DevExpress NuGet Gallery in Visual Studio to restore the NuGet packages used in this solution. See the following topic for more information: [Get Started with DevExpress Mobile UI for .NET MAUI](https://docs.devexpress.com/MAUI/403249/get-started).

   You can also refer to the following YouTube video for more information on how to get started with the DevExpress .NET MAUI Controls: [Setting up a .NET MAUI Project](https://www.youtube.com/watch?v=juJvl5UicIQ).

> * To run this example, you should have a [Universal](https://www.devexpress.com/subscriptions/universal.xml), [DXperience](https://www.devexpress.com/subscriptions/dxperience.xml), or [Office File API](https://www.devexpress.com/products/net/office-file-api/) subscription.  

Reviewer: ```suggestion
* To run this example, you need to own license to the DevExpress [Universal](https://www.devexpress.com/subscriptions/universal.xml), [DXperience](https://www.devexpress.com/subscriptions/dxperience.xml), or [Office File API](https://www.devexpress.com/products/net/office-file-api/) subscription.  
```

> * An email client on your device should support an email's HTML formatting. This project is tested on Outlook application for Android and iOS.

Reviewer: ```suggestion
* The email client on your device should support HTML formatting. We tested the project with Outlook for Android and iOS.
```

## Implementation Details

> * To implement mail merge, the application uses the [RichEditDocumentServer](https://docs.devexpress.com/OfficeFileAPI/DevExpress.XtraRichEdit.RichEditDocumentServer) class. Follow the steps below to load an email template and define a source of a recipient's data:

Reviewer: I don't want to use nested lists. Why don't we introduce section headers instead? 

```suggestion

### Implement Mail Merge

Use the [RichEditDocumentServer](https://docs.devexpress.com/OfficeFileAPI/DevExpress.XtraRichEdit.RichEditDocumentServer) class to load an email template and define a data source with recipients:
```

>   1. Copy files from the application bundle to the _AppData_ folder to access these files from the application code:

Reviewer: ```suggestion
  1. Copy files from the application bundle to the _AppData_ folder to access these files from code:
```

  
  ```csharp
  public async Task CopyWorkingFilesToAppData(string fileName) {
      using Stream fileStream = await FileSystem.Current.OpenAppPackageFileAsync(fileName);
      string targetFile = Path.Combine(FileSystem.Current.AppDataDirectory, fileName);
      using FileStream outputStream = File.OpenWrite(targetFile);
      fileStream.CopyTo(outputStream);
  }
  ```
  
  1. Call the [RichEditDocumentServerExtensions.LoadDocumentAsync](https://docs.devexpress.com/OfficeFileAPI/DevExpress.XtraRichEdit.RichEditDocumentServerExtensions.LoadDocumentAsync.overloads?p=netstandard) method to load an email template:
  
  ```csharp
  async Task<RichEditDocumentServer> MailMergeAsync(EmailTemplate emailTemplate) {
      string workingFilePath = Path.Combine(FileSystem.Current.AppDataDirectory, emailTemplate.DocumentSourcePath);
      RichEditDocumentServer mergeRichProcessor = new RichEditDocumentServer();
      await mergeRichProcessor.LoadDocumentAsync(workingFilePath);
      // ...
  }
  ```
  

>   1. Specify the [RichEditMailMergeOptions.DataSource](https://docs.devexpress.com/OfficeFileAPI/DevExpress.XtraRichEdit.RichEditMailMergeOptions.DataSource?p=netstandard) property to define a source of a recipient's data.

Reviewer: ```suggestion
  1. Assign the recipient data source to the [RichEditMailMergeOptions.DataSource](https://docs.devexpress.com/OfficeFileAPI/DevExpress.XtraRichEdit.RichEditMailMergeOptions.DataSource?p=netstandard) property.
```

  
  ```csharp
  async Task<RichEditDocumentServer> MailMergeAsync(EmailTemplate emailTemplate) {
      // ...
      mergeRichProcessor.Document.Fields[0].Locked = true;
      mergeRichProcessor.Options.MailMerge.DataSource = new List<object> { new { RecipientName = mailToCustomer.FirstName, SenderName = currentUserName} };
      // ...
  }
  ```
  

>   1. Call the [RichEditDocumentServer.MailMerge](https://docs.devexpress.com/OfficeFileAPI/DevExpress.XtraRichEdit.RichEditDocumentServer.MailMerge(DevExpress.XtraRichEdit.API.Native.Document)?p=netstandard) method to merge data to the current document and send the result to the specified [Document](https://docs.devexpress.com/OfficeFileAPI/DevExpress.XtraRichEdit.API.Native.Document?p=netstandard) instance.

Reviewer: ```suggestion
  1. Call the [RichEditDocumentServer.MailMerge](https://docs.devexpress.com/OfficeFileAPI/DevExpress.XtraRichEdit.RichEditDocumentServer.MailMerge(DevExpress.XtraRichEdit.API.Native.Document)?p=netstandard) method to merge data and send the result to the specified [Document](https://docs.devexpress.com/OfficeFileAPI/DevExpress.XtraRichEdit.API.Native.Document?p=netstandard) instance.
```

  
  ```csharp
  async Task<RichEditDocumentServer> MailMergeAsync(EmailTemplate emailTemplate) {
      // ...
      MailMergeOptions myMergeOptions = mergeRichProcessor.Document.CreateMailMergeOptions();
      RichEditDocumentServer resultDocumentProcessor = new RichEditDocumentServer();
      resultDocumentProcessor.CreateNewDocument();
      myMergeOptions.MergeMode = MergeMode.NewSection;
      mergeRichProcessor.MailMerge(resultDocumentProcessor.Document);
      return resultDocumentProcessor;
  }
  ```

> * You can save a mail merge result as a file, stream, or document. This application saves a mail merge as a document and calls the [SubDocument.GetHtmlText](https://docs.devexpress.com/OfficeFileAPI/DevExpress.XtraRichEdit.API.Native.SubDocument.GetHtmlText(DevExpress.XtraRichEdit.API.Native.DocumentRange-DevExpress.Office.Services.IUriProvider)?p=netstandard) method to convert mail merge result to HTML. This allows you to specify an email's custom HTML formatting. 

Reviewer: ```suggestion
### Save Document as HTML

Save the mail merge result as a document. Call the [SubDocument.GetHtmlText](https://docs.devexpress.com/OfficeFileAPI/DevExpress.XtraRichEdit.API.Native.SubDocument.GetHtmlText(DevExpress.XtraRichEdit.API.Native.DocumentRange-DevExpress.Office.Services.IUriProvider)?p=netstandard) method to convert the content to HTML.
```

> * Call the [Email.ComposeAsync](https://learn.microsoft.com/en-us/dotnet/api/microsoft.maui.applicationmodel.communication.email.composeasync?view=net-maui-7.0) method to send an email. You should also specify the required attributes in the following files: 

Reviewer: ```suggestion
### Send Emails

Call the [Email.ComposeAsync](https://learn.microsoft.com/en-us/dotnet/api/microsoft.maui.applicationmodel.communication.email.composeasync?view=net-maui-7.0) method to send emails. You should specify required attributes in the following files: 
```

# Mail Merge

This example uses Mail Merge to send emails. Data placeholders in an email template obtain their values from database records. Mail Merge functionality is included in our [Office File API](https://www.devexpress.com/products/net/office-file-api/) subscription.

## Requirements

* Please register the DevExpress NuGet Gallery in Visual Studio to restore the NuGet packages used in this solution. See the following topic for more information: [Get Started with DevExpress Mobile UI for .NET MAUI](https://docs.devexpress.com/MAUI/403249/get-started).

   You can also refer to the following YouTube video for more information on how to get started with the DevExpress .NET MAUI Controls: [Setting up a .NET MAUI Project](https://www.youtube.com/watch?v=juJvl5UicIQ).

> * To run this example, you need to own license to the DevExpress [Universal](https://www.devexpress.com/subscriptions/universal.xml), [DXperience](https://www.devexpress.com/subscriptions/dxperience.xml), or [Office File API](https://www.devexpress.com/products/net/office-file-api/) subscription.  

Reviewer 2: ```suggestion
* To run this example, you need a lDevExpress [Universal](https://www.devexpress.com/subscriptions/universal.xml), [DXperience](https://www.devexpress.com/subscriptions/dxperience.xml), or [Office File API](https://www.devexpress.com/products/net/office-file-api/) subscription.  
```

Author: ```suggestion
* To run this example, you need a DevExpress [Universal](https://www.devexpress.com/subscriptions/universal.xml), [DXperience](https://www.devexpress.com/subscriptions/dxperience.xml), or [Office File API](https://www.devexpress.com/products/net/office-file-api/) subscription.  
```
